File: java/solutions/03-advanced/ex086_money_value_object/Money.java
package fewolearning.exercises.advanced.ex086_money_value_object;

import java.math.BigDecimal;
import java.util.Currency;
import java.util.Objects;

/*
Exercise 086 - Money value object (reference solution).
*/
public final class Money {
    private final BigDecimal amount;
    private final Currency currency;

    public Money(BigDecimal amount, Currency currency) {
        this.amount = amount;
        this.currency = currency;
    }

    public Money add(Money other) {
        if (!currency.equals(other.currency)) {
            throw new IllegalArgumentException(
                    "Cannot add different currencies: " + currency + " and " + other.currency);
        }
        return new Money(amount.add(other.amount), currency);
    }

    @Override
    public boolean equals(Object other) {
        if (this == other) {
            return true;
        }
        if (!(other instanceof Money money)) {
            return false;
        }
        return amount.compareTo(money.amount) == 0 && currency.equals(money.currency);
    }

    @Override
    public int hashCode() {
        return Objects.hash(amount.stripTrailingZeros(), currency);
    }
}

File: php/solutions/02-intermediate/ex050_typed_properties/Money.php
<?php

declare(strict_types=1);

namespace FeWoLearning\Exercises\Intermediate\Ex050TypedProperties;

final class Money
{
    private int $cents;
    private string $currency;

    public function __construct(int $cents, string $currency)
    {
        $this->cents = $cents;
        $this->currency = $currency;
    }

    public function add(int|self $amount): self
    {
        if ($amount instanceof self) {
            if ($amount->currency !== $this->currency) {
                throw new \InvalidArgumentException('Currency mismatch');
            }

            return new self($this->cents + $amount->cents, $this->currency);
        }

        return new self($this->cents + $amount, $this->currency);
    }

    public function cents(): int
    {
        return $this->cents;
    }
}
